### PR TITLE
Moved shading modes from renderers to materials

### PR DESCRIPTION
**Pull request by favreau**

### CODE REVIEW

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

**Comment by ppodhajski** [comment]
retest this please

File: brayns/common/CMakeLists.txt [hunk 1]
```diff
@@ -75,13 +75,13 @@ set(BRAYNSCOMMON_PUBLIC_HEADERS
   transferFunction/TransferFunction.h
   types.h
   utils/Utils.h
+  utils/ImageUtils.h
```
Comment (tribal-tec) [inline brayns/common/CMakeLists.txt]: should remain private

File: brayns/Brayns.cpp [hunk 2]
```diff
@@ -1066,4 +1047,4 @@ AbstractManipulator& Brayns::getCameraManipulator()
 {
     return _impl->getCameraManipulator();
 }
-}
+} // namespace brayns
```
Comment (tribal-tec) [inline brayns/Brayns.cpp]: ? different clang-format version why?

File: ospray_modules/optix/cuda/renderer/Phong.h [hunk 3]
```diff
@@ -263,7 +261,8 @@ static __device__ void phongShadowed(float3 p_Ko)
 static __device__ void phongShade(float3 p_Kd, float3 p_Ka, float3 p_Ks,
                                   float3 p_Kr, float3 p_Ko,
                                   float p_refractionIndex, float p_phong_exp,
-                                  float p_glossiness, float3 p_normal,
+                                  float p_glossiness,
+                                  unsigned int p_shadindMode, float3 p_normal,
```
Comment (karjonas) [inline ospray_modules/optix/cuda/renderer/Phong.h]: typo shadin(d)Mode

File: brayns/Brayns.cpp (hunk 2 above)
Comment (favreau) [reply brayns/Brayns.cpp]: hum... yes, I am using clang-format version 3.8.0-2ubuntu4... Which one are you using?

Comment (favreau) [reply brayns/Brayns.cpp]: hold on, this probably comes from my home PC where I am using Ubuntu 18.04, or maybe from my Mac... Let me put things back together from my Ubuntu machine.

File: engines/ospray/ispc/render/ExtendedOBJMaterial.ih [hunk 4]
```diff
@@ -28,6 +28,14 @@
 #include <ospray/SDK/texture/TextureParam.ih>
 #include <ospray/SDK/volume/Volume.ih>
 
+// Caution: This has to match the Brayns::MaterialShadingMode
+enum MaterialShadingMode
```
Comment (karjonas) [inline engines/ospray/ispc/render/ExtendedOBJMaterial.ih]: Can you not put this in a header file somewhere and include that in ispc and cpp code?

File: brayns/Brayns.cpp (hunk 2 above)
Comment (tribal-tec) [reply brayns/Brayns.cpp]: also 3.8 from Ubuntu 16.04. It must be a more recent version, yes.

File: engines/ospray/ispc/render/ExtendedOBJMaterial.ih (hunk 4 above)
Comment (favreau) [reply engines/ospray/ispc/render/ExtendedOBJMaterial.ih]: I tried, but could not find a way to do so... I am having another look now, with not much hope though.

**Comment by tribal-tec** [comment]
The `CMake/common` change probably also needs to be undone, as it looks like it reverts back one commit?

File: engines/ospray/ispc/render/ExtendedOBJMaterial.ih (hunk 4 above)
Comment (karjonas) [reply engines/ospray/ispc/render/ExtendedOBJMaterial.ih]: I did something similar in brayns/common/geometry/Cone.h

File: brayns/common/CMakeLists.txt [hunk 5]
```diff
@@ -81,7 +81,6 @@ set(BRAYNSCOMMON_PUBLIC_HEADERS
 )
 
 set(BRAYNSCOMMON_HEADERS
-  utils/ImageUtils.h
```
Comment (tribal-tec) [inline brayns/common/CMakeLists.txt]: undo delete?

File: tests/braynsTestData.cpp [hunk 6]
```diff
@@ -119,6 +120,24 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_color_and_compare)
     camera.setTarget(rotCenter);
     camera.setPosition(camPos + 0.9 * (rotCenter - camPos));
 
+    auto& scene = brayns.getEngine().getScene();
```
Comment (tribal-tec) [inline tests/braynsTestData.cpp]: that outlines a bit of the problem of this approach: the user now cannot 'easily' change the shading of the entire scene (in the UI not all anymore). Also, renderers which don't evaluate the material shading mode (like the basic renderer), shade anyway the way they want :)

**Comment by hernando** [comment]
What are the motivations for this change? I just read the commit name, but logically it doesn't make much sense to me.

**Comment by favreau** [comment]
The shading is a material property, not a renderer one. I understand this is more convenient from a user point of view to change everything in one click, but in reality, one needs more granularity. The bounding box is a good example of geometries that needs specific shading. To me, having this attribute at a renderer level is what does not make sense.

File: brayns/common/CMakeLists.txt (hunk 5 above)
Comment (favreau) [reply brayns/common/CMakeLists.txt]: oops, thanks.

**Comment by hernando** [comment]
I misread shading by shadowing. Maybe it makes sense after all.

**Comment by straaljager** [comment]
"The shading is a material property, not a renderer one." I think it would be good to discuss this type of design choice. As far as I know, all rendering engines on the market use the same shading (e.g. AO, path tracing, Phong) for all materials in the scene. I'm not showing we should be doing exactly the same as what other renderers out there are doing, as there might be good reasons to implement it differently from the traditional way. But that needs discussion, as currently it's unclear.

**Comment by straaljager** [comment]
Sounds like there is a bit of confusion about terminology. Disregard my previous comment.

**Comment by favreau** [comment]
No worries. Just to clarify, I am not dealing with rendering options there, this PR only is about the "color" of the surface according to the light incidence. It's somewhat like an extremely simplified BRDF implementation. We will eventually have to implement proper BRDF materials, and this will of course require discussions.

**Comment by straaljager** [comment]
OK, it all makes sense now. Looking forward to better materials and accompanying discussions :)

File: CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +9,7 @@ cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
 project(Brayns VERSION 0.8.0)
 set(Brayns_VERSION_ABI 1)
 
+set(BRAYNS_SOURCE_DIR ${PROJECT_SOURCE_DIR})
```
Comment (tribal-tec) [inline CMakeLists.txt]: ? Use `CMAKE_SOURCE_DIR` where you need to get this.

File: brayns/common/commonTypes.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#ifndef COMMONTYPES_H
```
Comment (tribal-tec) [inline brayns/common/commonTypes.h]: no copyright header here?

File: brayns/common/types.h [hunk 9]
```diff
@@ -30,6 +30,7 @@
 #define SERIALIZATION_FRIEND(type) \
     friend void staticjson::init(type*, staticjson::ObjectHandler*);
 
+#include "commonTypes.h"
```
Comment (tribal-tec) [inline brayns/common/types.h]: should be system include like the one below

File: tests/braynsTestData.cpp [hunk 10]
```diff
@@ -119,6 +120,24 @@ BOOST_AUTO_TEST_CASE(render_circuit_with_color_and_compare)
...
+    for (size_t i = 0; i < scene.getNumModels(); ++i)
+    {
+        auto modelDescriptor = scene.getModel(i);
```
Comment (tribal-tec) [inline tests/braynsTestData.cpp]: this is wrong btw, as `getModel` needs a modelID, not an index.

**Comment by tribal-tec** [comment]
All the test images should not change, right? Maybe just revert of all them?

**Pull request closed without merging**